[aot_autograd] Add support for tracking and handling opaque objects.

aorenste opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #171137

Summary:

This commit adds comprehensive support for tracking and handling opaque objects
(non-tensor, non-scriptable objects) through FX tracing and AOTAutograd.

Key changes:

1. **Proxy tracking infrastructure**:
   - When we see opaque objects use the script_object_tracker to track them.

2. **Opaque input descriptors**:
   - Add `opaque_inp_descs` field to `ViewAndMutationMeta` to store descriptors for
     opaque objects synthesized from subclass inputs (e.g., ProcessGroups from DTensors)
   - These are tracked separately from `subclass_inp_meta` since they represent
     non-tensor objects

3. **Runtime evaluation**:
   - Add `_evaluate_opaque_descriptor()` in runtime_wrappers to evaluate descriptors
     at runtime and extract actual opaque values
   - Supports nested descriptors like `SubclassGetAttrAOTInput`

cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

aorenste committed:
[aot_autograd] Add support for tracking and handling opaque objects.

Summary:

This commit adds comprehensive support for tracking and handling opaque objects
(non-tensor, non-scriptable objects) through FX tracing and AOTAutograd.

Key changes:

1. **Proxy tracking infrastructure**:
   - Add `opaque_tracker` to `_ProxyTracer` and `ProxyTorchDispatchMode` to track
     proxy associations for opaque objects, similar to existing tensor/symint tracking
   - Extend `set_proxy_slot()` and `get_proxy_slot()` to handle opaque values

2. **Opaque input descriptors**:
   - Add `opaque_inp_descs` field to `ViewAndMutationMeta` to store descriptors for
     opaque objects synthesized from subclass inputs (e.g., ProcessGroups from DTensors)
   - These are tracked separately from `subclass_inp_meta` since they represent
     non-tensor objects

3. **Runtime evaluation**:
   - Add `_evaluate_opaque_descriptor()` in runtime_wrappers to evaluate descriptors
     at runtime and extract actual opaque values
   - Supports nested descriptors like `SubclassGetAttrAOTInput`

[ghstack-poisoned]

aorenste committed:
Update on "[aot_autograd] Add support for tracking and handling opaque objects."


Summary:

This commit adds comprehensive support for tracking and handling opaque objects
(non-tensor, non-scriptable objects) through FX tracing and AOTAutograd.

Key changes:

1. **Proxy tracking infrastructure**:
   - Add `opaque_tracker` to `_ProxyTracer` and `ProxyTorchDispatchMode` to track
     proxy associations for opaque objects, similar to existing tensor/symint tracking
   - Extend `set_proxy_slot()` and `get_proxy_slot()` to handle opaque values

2. **Opaque input descriptors**:
   - Add `opaque_inp_descs` field to `ViewAndMutationMeta` to store descriptors for
     opaque objects synthesized from subclass inputs (e.g., ProcessGroups from DTensors)
   - These are tracked separately from `subclass_inp_meta` since they represent
     non-tensor objects

3. **Runtime evaluation**:
   - Add `_evaluate_opaque_descriptor()` in runtime_wrappers to evaluate descriptors
     at runtime and extract actual opaque values
   - Supports nested descriptors like `SubclassGetAttrAOTInput`

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

angelayi commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -283,6 +288,10 @@ def set_proxy_slot(  # type: ignore[no-redef]
         # We DO want to clobber proxies, with a similar rationale as for tensors.
         assert isinstance(proxy, Proxy)
         tracer.script_object_tracker[obj] = proxy
+    elif is_opaque_value(obj):
```
Sorry, do you happen to have a test case that runs into the case where these changes are needed?

angelayi replied:
Sorry ignore the comment ab updating the test -- instead of creating a new tracker for opaque objects can we combine it with the ScriptObject type code? When we trace symbolically we convert the opaque objects to FakeScriptObjects, so when we trace with real I think we can also just use the same script object trackers. I think the change you need to make is just replace all `isinstance(obj, _AnyScriptObject)` with `isinstance(obj, _AnyScriptObject) or is_opaque_value(obj)`

aorenste committed:
Update on "[aot_autograd] Add support for tracking and handling opaque objects."


Summary:

This commit adds comprehensive support for tracking and handling opaque objects
(non-tensor, non-scriptable objects) through FX tracing and AOTAutograd.

Key changes:

1. **Proxy tracking infrastructure**:
   - Add `opaque_tracker` to `_ProxyTracer` and `ProxyTorchDispatchMode` to track
     proxy associations for opaque objects, similar to existing tensor/symint tracking
   - Extend `set_proxy_slot()` and `get_proxy_slot()` to handle opaque values

2. **Opaque input descriptors**:
   - Add `opaque_inp_descs` field to `ViewAndMutationMeta` to store descriptors for
     opaque objects synthesized from subclass inputs (e.g., ProcessGroups from DTensors)
   - These are tracked separately from `subclass_inp_meta` since they represent
     non-tensor objects

3. **Runtime evaluation**:
   - Add `_evaluate_opaque_descriptor()` in runtime_wrappers to evaluate descriptors
     at runtime and extract actual opaque values
   - Supports nested descriptors like `SubclassGetAttrAOTInput`

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

aorenste committed:
Update on "[aot_autograd] Add support for tracking and handling opaque objects."


Summary:

This commit adds comprehensive support for tracking and handling opaque objects
(non-tensor, non-scriptable objects) through FX tracing and AOTAutograd.

Key changes:

1. **Proxy tracking infrastructure**:
   - Add `opaque_tracker` to `_ProxyTracer` and `ProxyTorchDispatchMode` to track
     proxy associations for opaque objects, similar to existing tensor/symint tracking
   - Extend `set_proxy_slot()` and `get_proxy_slot()` to handle opaque values

2. **Opaque input descriptors**:
   - Add `opaque_inp_descs` field to `ViewAndMutationMeta` to store descriptors for
     opaque objects synthesized from subclass inputs (e.g., ProcessGroups from DTensors)
   - These are tracked separately from `subclass_inp_meta` since they represent
     non-tensor objects

3. **Runtime evaluation**:
   - Add `_evaluate_opaque_descriptor()` in runtime_wrappers to evaluate descriptors
     at runtime and extract actual opaque values
   - Supports nested descriptors like `SubclassGetAttrAOTInput`

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

aorenste commented on torch/_functorch/_aot_autograd/graph_capture_wrappers.py:
```diff
@@ -1201,12 +1201,38 @@ def aot_dispatch_subclass(
+                primals, tangents = args
+                tensor_primals, opaque_args = (
+                    primals[:-num_opaque],
+                    primals[-num_opaque:],
```
@angelayi Ok - test added and changed wrap_tensor_subclasses_maybe_joint (and friends) to understand opaque args so we don't need this particular hack.

aorenste committed:
Update on "[aot_autograd] Add support for tracking and handling opaque objects."


Summary:

This commit adds comprehensive support for tracking and handling opaque objects
(non-tensor, non-scriptable objects) through FX tracing and AOTAutograd.

Key changes:

1. **Proxy tracking infrastructure**:
   - Add `opaque_tracker` to `_ProxyTracer` and `ProxyTorchDispatchMode` to track
     proxy associations for opaque objects, similar to existing tensor/symint tracking
   - Extend `set_proxy_slot()` and `get_proxy_slot()` to handle opaque values

2. **Opaque input descriptors**:
   - Add `opaque_inp_descs` field to `ViewAndMutationMeta` to store descriptors for
     opaque objects synthesized from subclass inputs (e.g., ProcessGroups from DTensors)
   - These are tracked separately from `subclass_inp_meta` since they represent
     non-tensor objects

3. **Runtime evaluation**:
   - Add `_evaluate_opaque_descriptor()` in runtime_wrappers to evaluate descriptors
     at runtime and extract actual opaque values
   - Supports nested descriptors like `SubclassGetAttrAOTInput`

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

angelayi reviewed (approved):
thank you!

aorenste committed:
Update on "[aot_autograd] Add support for tracking and handling opaque objects."


Summary:

This commit adds comprehensive support for tracking and handling opaque objects
(non-tensor, non-scriptable objects) through FX tracing and AOTAutograd.

Key changes:

1. **Proxy tracking infrastructure**:
   - When we see opaque objects use the script_object_tracker to track them.

2. **Opaque input descriptors**:
   - Add `opaque_inp_descs` field to `ViewAndMutationMeta` to store descriptors for
     opaque objects synthesized from subclass inputs (e.g., ProcessGroups from DTensors)
   - These are tracked separately from `subclass_inp_meta` since they represent
     non-tensor objects

3. **Runtime evaluation**:
   - Add `_evaluate_opaque_descriptor()` in runtime_wrappers to evaluate descriptors
     at runtime and extract actual opaque values
   - Supports nested descriptors like `SubclassGetAttrAOTInput`

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

aorenste committed:
Update on "[aot_autograd] Add support for tracking and handling opaque objects."


Summary:

This commit adds comprehensive support for tracking and handling opaque objects
(non-tensor, non-scriptable objects) through FX tracing and AOTAutograd.

Key changes:

1. **Proxy tracking infrastructure**:
   - When we see opaque objects use the script_object_tracker to track them.

2. **Opaque input descriptors**:
   - Add `opaque_inp_descs` field to `ViewAndMutationMeta` to store descriptors for
     opaque objects synthesized from subclass inputs (e.g., ProcessGroups from DTensors)
   - These are tracked separately from `subclass_inp_meta` since they represent
     non-tensor objects

3. **Runtime evaluation**:
   - Add `_evaluate_opaque_descriptor()` in runtime_wrappers to evaluate descriptors
     at runtime and extract actual opaque values
   - Supports nested descriptors like `SubclassGetAttrAOTInput`

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste commented:
BC Failure is:
```
[warning]Function get_proxy_slot: obj changed from Union[Tensor, _AnyScriptObjectType, PySymType] to Union[Tensor, _AnyScriptObjectType, PySymType, OpaqueType]
```
adding a new type to the union is fine.
@pytorchbot merge -i
